api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,600 @@
 package aws
 
-import "fmt"
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"sort"
+	"strconv"
+
+	"github.com/zpatrick/go-bytesize"
+	cache "github.com/zpatrick/go-cache"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/quintilesims/layer0/api/job"
+	"github.com/quintilesims/layer0/api/provider"
+	"github.com/quintilesims/layer0/api/scaler"
+	awsc "github.com/quintilesims/layer0/common/aws"
+	"github.com/quintilesims/layer0/common/config"
+	"github.com/quintilesims/layer0/common/errors"
+	"github.com/quintilesims/layer0/common/models"
+)
 
 type EnvironmentScaler struct {
+	Client              *awsc.Client
+	EnvironmentProvider provider.EnvironmentProvider
+	ServiceProvider     provider.ServiceProvider
+	TaskProvider        provider.TaskProvider
+	JobStore            job.Store
+	Config              config.APIConfig
+	deployCache         *cache.Cache
+	environmentCache    *cache.Cache
 }
 
-func NewEnvironmentScaler() *EnvironmentScaler {
-	return &EnvironmentScaler{}
+func NewEnvironmentScaler(a *awsc.Client, e provider.EnvironmentProvider, s provider.ServiceProvider, t provider.TaskProvider, j job.Store, c config.APIConfig) *EnvironmentScaler {
+	return &EnvironmentScaler{
+		Client:              a,
+		EnvironmentProvider: e,
+		ServiceProvider:     s,
+		TaskProvider:        t,
+		JobStore:            j,
+		Config:              c,
+		deployCache:         cache.New(),
+		environmentCache:    cache.New(),
+	}
 }
 
+type ProviderDistribution struct {
+	Errs      []error
+	Providers []*scaler.ResourceProvider
+}
+
+// Scale determines whether or not instances need to be added to or removed from a Layer0 environment, and makes any necessary changes.
+// It consists of three primary logical groupings:
+// 1. Gather all providers (instances) and consumers (tasks/services) of resources.
+// 2. Calculate the optimal distribution of consumers among providers, including whether instances should be added or removed.
+// 3. Update the AutoScaling Group to realize the calculated changes.
 func (e *EnvironmentScaler) Scale(environmentID string) error {
-	return fmt.Errorf("EnvironmentScaler not implemented")
+	clusterName := addLayer0Prefix(e.Config.Instance(), environmentID)
+
+	resourceConsumers, resourceProviders, err := e.GetCurrentState(clusterName)
+	if err != nil {
+		return err
+	}
+
+	var incompatibilityError error
+	resourceProviders, unusedProviders, desiredScale, err := e.CalculateOptimizedState(clusterName, resourceConsumers, resourceProviders)
+
+	if err != nil {
+		if err, ok := err.(*errors.ServerError); ok && err.Code == errors.IncompatibleConsumerAndProvider {
+			incompatibilityError = err
+		} else {
+			return err
+		}
+	}
+
+	if err := e.ScaleToState(clusterName, desiredScale, unusedProviders); err != nil {
+		return err
+	}
+
+	return incompatibilityError
```

Comment:
So the intention is that if you have an incompatibilityError:
 1. you log the error
 2. you will continue to attempt to scale disregarding the consumer that requires too many resources
 3. bubble it up to the caller of Scale

My question is about point 3. What is the caller of scale expected to do with that error apart from logging and moving on? Or to put it another way, what is the caller of scale supposed to do with that error?

Reply:
The caller of Scale treats it like any other error that would bubble up from `Scale()`, and returns it to the user as an error. It's still a problem that a user needs to know about and deal with, so I return it as an error. If it were _only_ logged to the API's stdout, the only way a user would learn about the problem is if they were running the API locally (or had SSH'd into the API's instance and inspected the logs).

If you have a better idea for handling it, I'd love to hear it!

Reply:
> and returns it to the user as an error.

The call to this method is initiated by both the scaler and a direct action from the user (i.e., run task or create service). I was more wondering about what the scaler intends to do with the error other than just logging it again. But it is a minor point though so I say it is fine as it is.